autotx/AutoTx.py

```diff
@@ -1,73 +1,98 @@
+        self, manager: SafeManager, network: NetworkInfo, agent_factories: list[Callable[['AutoTx', UserProxyAgent, Optional[Dict[str, Any]]], Agent]], config: Optional[Config],
+        get_llm_config: Callable[[], Optional[Dict[str, Any]]]
     ):
         self.manager = manager
         self.network = network
+        self.get_llm_config = get_llm_config
         if config:
             self.config = config
-        self.agents = [factory(self) for factory in agent_factories]
+        self.agent_factories = agent_factories
 
     def run(self, prompt: str, non_interactive: bool):
-        print(f"Defining goal for prompt: '{prompt}'")
-       
-        agents_information = self.get_agents_information()
+        print("Running AutoTx with the following prompt: ", prompt)
+
+        user_proxy = UserProxyAgent(
+            name="user_proxy",
+            is_termination_msg=lambda x: x.get("content", "") and x.get("content", "").rstrip().endswith("TERMINATE"),
+            human_input_mode="NEVER",
+            max_consecutive_auto_reply=20,
+            system_message=f"You are a user proxy. You will be interacting with the agents to accomplish the tasks.",
+            llm_config=self.get_llm_config(),
+            code_execution_config=False,
+        )
+        agents = [factory(self, user_proxy, self.get_llm_config()) for factory in self.agent_factories]
+
+        agents_information = self.get_agents_information(agents)
 
         goal = build_goal(prompt, agents_information, self.manager.address, non_interactive)
 
-        print(f"Defining tasks for goal: '{goal}'")
-        tasks: list[Task] = define_tasks(goal, agents_information, self.agents)
-   
-        self.run_for_tasks(tasks, non_interactive)
+        verifier_agent = AssistantAgent(
+            name="verifier",
+            is_termination_msg=lambda x: x.get("content", "") and x.get("content", "").rstrip().endswith("TERMINATE"),
+            system_message=dedent(
+                    """
+                    You are an expert in verifiying if user goals are met.
+                    You analyze chat and respond with TERMINATE if the goal is met.
+                    You can consider the goal met if the other agents have prepared the necessary transactions.
+                    """
+                ),
+            llm_config=self.get_llm_config(),
+            human_input_mode="NEVER",
+            code_execution_config=False,
+        )
 
-    def run_for_tasks(self, tasks: list[Task], non_interactive: bool):
-        print(f"Running tasks...")
-        Crew(
-            agents=self.agents,
-            tasks=tasks,
-            verbose=self.config.verbose,
-            process=Process.sequential,
-            function_calling_llm=open_ai_llm,
-        ).kickoff()
+        groupchat = autogen.GroupChat(agents=[agent.autogen_agent for agent in agents] + [user_proxy, verifier_agent], messages=[], max_round=12)
```

Comment:
What happens when `max_rounds` is reached? Does it simply exit?

Reply:
Yes